### PR TITLE
Add 64bit indexing support for softmax

### LINKED ISSUE #52715: CUDA Illegal memory access for softmax
Related https://github.com/pytorch/pytorch/issues/52716

```python
import torch
x = torch.randn((1100000000, 2), device="cuda:0", dtype=torch.float16, requires_grad=True)
y = torch.nn.functional.log_softmax(x, dim=-1)
y.backward(y)
torch.cuda.synchronize()
```
```
Traceback (most recent call last):
  File "/home/gaoxiang/misc/softmax.py", line 6, in <module>
    torch.cuda.synchronize()
  File "/home/gaoxiang/.local/lib/python3.9/site-packages/torch/cuda/__init__.py", line 402, in synchronize
    return torch._C._cuda_synchronize()
RuntimeError: CUDA error: an illegal memory access was encountered
Exception raised from device_synchronize at ../c10/cuda/CUDAFunctions.cpp:135 (most recent call first):
frame #0: <unknown function> + 0x8101f (0x7fd7fefe201f in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x80d85 (0x7fd7fefe1d85 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #2: <unknown function> + 0x80b49 (0x7fd7fefe1b49 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #3: std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > ()>::operator()() const + 0x4c (0x7fd82bf16102 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #4: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7fd7fefe0dc0 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #5: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x7b (0x7fd7fefdf58c in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #6: c10::cuda::device_synchronize() + 0x7c (0x7fd7ff04e27b in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libc10_cuda.so)
frame #7: THCPModule_cudaSynchronize(_object*, _object*) + 0x3f (0x7fd838be3dec in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_python.so)
<omitting python frames>
frame #21: __libc_start_main + 0xd5 (0x7fd8659aeb25 in /usr/lib/libc.so.6)
```

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @ngimel

### LINKED ISSUE #52716: CUDA error: invalid configuration argument for softmax
Related: https://github.com/pytorch/pytorch/issues/52715
```python
import torch
x = torch.randn((2200000000, 1), device="cuda:0", dtype=torch.float16, requires_grad=True)
torch.cuda.profiler.start()
y = torch.nn.functional.log_softmax(x, dim=-1)
y.backward(y)
torch.cuda.profiler.stop()
torch.cuda.synchronize()
```
```
Traceback (most recent call last):
  File "/home/gaoxiang/misc/softmax.py", line 4, in <module>
    y = torch.nn.functional.log_softmax(x, dim=-1)
  File "/home/gaoxiang/.local/lib/python3.9/site-packages/torch/nn/functional.py", line 1672, in log_softmax
    ret = input.log_softmax(dim)
RuntimeError: CUDA error: invalid configuration argument
Exception raised from dispatch_softmax_forward at /home/gaoxiang/pytorch-cmake/aten/src/ATen/native/cuda/PersistentSoftmax.cuh:268 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x5c (0x7f16cef45e3c in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x64 (0x7f16cef19828 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #2: void dispatch_softmax_forward<c10::Half, c10::Half, float, true>(c10::Half*, c10::Half const*, int, int, int) + 0x2ef (0x7f16d0aa249f in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #3: <unknown function> + 0x1b3c2dd (0x7f16d0aa52dd in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #4: at::Tensor at::native::(anonymous namespace)::host_softmax<at::native::(anonymous namespace)::LogSoftMaxForwardEpilogue, true>(at::Tensor const&, long, bool) + 0x485 (0x7f16d0aa7455 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #5: at::native::log_softmax_cuda(at::Tensor const&, long, bool) + 0x21 (0x7f16d0a9f0e1 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #6: <unknown function> + 0x2608468 (0x7f16d1571468 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #7: <unknown function> + 0x2608518 (0x7f16d1571518 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #8: <unknown function> + 0x153562a (0x7f16f56b062a in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #9: <unknown function> + 0x1536289 (0x7f16f56b1289 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #10: <unknown function> + 0x2c9e5e6 (0x7f16f6e195e6 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #11: <unknown function> + 0x2c9e75b (0x7f16f6e1975b in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #12: at::_log_softmax(at::Tensor const&, long, bool) + 0x13c (0x7f16f53a15ac in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #13: at::native::log_softmax(at::Tensor const&, long, c10::optional<c10::ScalarType>) + 0x75 (0x7f16f50cbb95 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #14: <unknown function> + 0x17e95f8 (0x7f16f59645f8 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #15: <unknown function> + 0x17e9697 (0x7f16f5964697 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #16: at::Tensor::log_softmax(long, c10::optional<c10::ScalarType>) const + 0x137 (0x7f16f5acba97 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #17: <unknown function> + 0x55c212 (0x7f16fc960212 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_python.so)
<omitting python frames>
frame #32: __libc_start_main + 0xd5 (0x7f1729217b25 in /usr/lib/libc.so.6)
```

cc @ngimel

### PR DESCRIPTION
**Pull request by zasdfgbnm**
fixes https://github.com/pytorch/pytorch/issues/52715 https://github.com/pytorch/pytorch/issues/52716

split across batch dimension

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
cc: @ptrblck

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 1]
```diff
@@ -712,8 +712,17 @@ Tensor host_softmax(const Tensor & input_, const int64_t dim_, const bool half_t
...
-                output.data_ptr<scalar_t>(), input.data_ptr<scalar_t>(), dim_size, dim_size, outer_size);
+            auto output_ptr = output.data_ptr<scalar_t>();
+            auto input_ptr = input.data_ptr<scalar_t>();
+            int64_t remaining = outer_size;
+            int64_t chunk_size = (1<<31 - 1) / dim_size;
+            while(remaining > 0) {
+              dispatch_softmax_forward<scalar_t, scalar_t, accscalar_t, is_log_softmax>(
+                output_ptr, input_ptr, dim_size, dim_size, std::min<int64_t>(remaining, chunk_size));
+              input_ptr += chunk_size * dim_size;
+              output_ptr += chunk_size;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/SoftMax.cu:723]: why is output advanced by a different offset? input and output sizes are the same for softmax

**Review by ngimel — COMMENTED** [review]
Can you please add tests?

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 1 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/SoftMax.cu:723]: This is a bug. I was thinking about `logsumexp` when writing this line, which was wrong.

**Comment by zasdfgbnm** [comment]
@ngimel I have fixed the bug you catch,  and added a test. The test passes on my 3090.

File: test/test_nn.py [hunk 2]
```diff
@@ -11975,6 +11975,25 @@ def test_softmax_results(self, device, dtype):
                         self.assertEqual(grad_input, ref_grad_input)
                         self.assertEqual(input.grad, ref_input.grad)
 
+    @onlyCUDA
+    @dtypesIfCUDA(torch.float, torch.half)
+    @largeTensorTest("20GB")
```
Comment (zasdfgbnm) [inline test/test_nn.py:11980]: On my 3090, half takes ~18GB mem, and float takes ~19.8GB

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 1 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/SoftMax.cu]: fixed

File: test/test_nn.py (hunk 2 above)
Comment (ngimel) [reply test/test_nn.py:11980]: Will these tests run in your CI?

**Review by ngimel — APPROVED** [review]

File: test/test_nn.py (hunk 2 above)
Comment (zasdfgbnm) [reply test/test_nn.py:11980]: Our CI has A100 and 3090, so yes!

**Pull request closed without merging**